### PR TITLE
DYN-7391: As a Dynamo dev, I want to update ASM component support for Global Launch

### PR DESCRIPTION
**Pull request by sm6srw**
### Purpose

This Pull Request does:
* add support for Autodesk Shared Components (ASC) to the Install Detective.
* add unit tests covering the new ASC wrapper functionality
* add support for ASM 231

Notes:
* The install detective is now looking for files both in installed host products and installed ASC components. We will eventually stop looking in installed host products and rely only on ASC.
* Dynamo still supports ASM 230.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Added support for ASM 231 and Autodesk Shared Components in preparation for Global launch

### CODE REVIEW

File: src/DynamoCore/DynamoCore.csproj [hunk 1]
```diff
@@ -112,15 +113,16 @@
...
     <PropertyGroup>
-      <LibGProtoGeometryLibGLocale>$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\libg_locale</LibGProtoGeometryLibGLocale>
+      <LibGProtoGeometryLibGLocale>$(PkgDynamoVisualProgramming_LibG_231_0_0)\tools\libg_locale</LibGProtoGeometryLibGLocale>
     </PropertyGroup>
     <ItemGroup>
-        <LibGInterface Include="$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\netstandard2.0\$(LibGOsToken)\LibG.Interface.dll" />
-        <LibGProtoGeometryDLL Include="$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\netstandard2.0\$(LibGOsToken)\ProtoGeometry.dll" />
-        <LibGProtoGeometryXML Include="$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\netstandard2.0\$(LibGOsToken)\ProtoGeometry.XML" />
-        <LibG230Deps Include="$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\netstandard2.0\$(LibGOsToken)\asm_deps.csproj" />
-        <LibGProtoGeometryUICulture Include="$(PkgDynamoVisualProgramming_LibG_230_0_0)\tools\netstandard2.0\$(LibGOsToken)\$(UICulture)\*" />
```
Comment (QilongTang) [inline src/DynamoCore/DynamoCore.csproj:122]: Just curious why the 230 related ones needs to be deleted?

File: src/Tools/DynamoInstallDetective/AscSDKWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        /// <summary>
+        /// Get the major version of all ASC packages installed on the local machine
+        /// </summary>
+        /// <returns>An array of major versions</returns>
+        public static string[] GetMajorVersions()
+        {
+            string[] majorVersions = [];
+            var baseKey = Registry.LocalMachine;
+            var subkey = baseKey.OpenSubKey(@"SOFTWARE\Autodesk\SharedComponents");
```
Comment (QilongTang) [inline src/Tools/DynamoInstallDetective/AscSDKWrapper.cs:145]: Can we use `registryKey` here?

File: src/Tools/DynamoInstallDetective/ProductLookUp.cs [hunk 3]
```diff
@@ -348,6 +348,41 @@ public virtual Tuple<int, int, int, int> GetVersionInfoFromFile(string filePath)
         }
     }
 
+#if NET6_0_OR_GREATER
+    [SupportedOSPlatform("windows")]
+#endif
+    public class InstalledAscLookUp : InstalledProductLookUp
+    {
+
+        public InstalledAscLookUp(string fileLookup) : base(@"Autodesk Shared Components", fileLookup)
```
Comment (QilongTang) [inline src/Tools/DynamoInstallDetective/ProductLookUp.cs:357]: Should we define `@"Autodesk Shared Components"` as a const string?

File: src/Tools/DynamoInstallDetective/AscSDKWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            }
+
+            var status = ReadASCVersionFromEnv() == ASC_STATUS.SUCCESS ? ASC_STATUS.SUCCESS : ReadASCInstallPathFromRegistry(majorRelease);
+            installedPath = installPath;
+
+            return status;
+        }
+
+        /// <summary>
+        /// Get the major version of all ASC packages installed on the local machine
```
Comment (QilongTang) [inline src/Tools/DynamoInstallDetective/AscSDKWrapper.cs:138]: Are these more like version informations including the package/ component names and other properties?

File: src/Tools/DynamoInstallDetective/ProductLookUp.cs [hunk 5]
```diff
@@ -348,6 +348,41 @@ public virtual Tuple<int, int, int, int> GetVersionInfoFromFile(string filePath)
         }
     }
 
+#if NET6_0_OR_GREATER
+    [SupportedOSPlatform("windows")]
+#endif
+    public class InstalledAscLookUp : InstalledProductLookUp
```
Comment (QilongTang) [inline src/Tools/DynamoInstallDetective/ProductLookUp.cs:354]: Would you add some comments to the purpose of this class?

**Review by QilongTang — APPROVED** [review]
LGTM with some comments. Please confirm, for a case where both ASC components and ASM binaries exist for a particular product, the ASM Dynamo decides to load is determined by the latest version or sequence?

File: src/DynamoCore/DynamoCore.csproj (hunk 1 above)
Comment (sm6srw) [reply src/DynamoCore/DynamoCore.csproj:122]: We always link against the latest version of LIBG. But we load the one that corresponds to the ASM version found.

File: src/Tools/DynamoInstallDetective/AscSDKWrapper.cs (hunk 2 above)
Comment (sm6srw) [reply src/Tools/DynamoInstallDetective/AscSDKWrapper.cs:145]: For convenience, this method is static.

**Comment by sm6srw** [comment]
> LGTM with some comments. Please confirm, for a case where both ASC components and ASM binaries exist for a particular product, the ASM Dynamo decides to load is determined by the latest version or sequence?

I have added more comments. Take a look and let me know if it is clearer now.

**Review by QilongTang — APPROVED** [review]

**Comment by QilongTang** [comment]
Rerunning unit tests on https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16494/

**Comment by sm6srw** [comment]
Passed!

**Pull request merged**